### PR TITLE
MPP-3328 & MPP-3327 - Premium Upsell

### PR DESCRIPTION
**Pull request by lloan**
This PR fixes MPP-3328 and MPP-3327 

# New feature description
- MPP-3327 5 Mask Premium Upsell Header 
- Show relay metrics header to free users

# Screenshot (if applicable)

image

# How to test

MPP-3327

When a user hits their 5 mask limit take over the top part of the relay experience wit the following header. The header cannot be dismissed.

Do not show any other banners at the top of the experience (including if they don’t have the add-on)

CSAT banner can stay

Upgrade to premium link should be instrumented in GA (e.g., “upgrade-premium-header-mask-limit”)

Upgrade to premium CTA takes user to premium price plans 

Designs:

US:[How to upsell 5 mask limit](https://www.figma.com/file/i0SNijb8Twcn7czbzcDLYj/How-to-upsell-5-mask-limit?type=design&node-id=337-7972&mode=design&t=ABRogTnrkrIIUdMH-4) 

Non-US: [How to upsell 5 mask limit](https://www.figma.com/file/i0SNijb8Twcn7czbzcDLYj/How-to-upsell-5-mask-limit?type=design&node-id=513-9586&mode=design&t=qwstSpGIsDdEs5vE-4)

MPP-3328
Show free users a similar welcome and metrics like premium users see that include (email masks used, emails blocked, emails forwarded, and trackers removed)

US/Canada version: [How to upsell 5 mask limit](https://www.figma.com/file/i0SNijb8Twcn7czbzcDLYj/How-to-upsell-5-mask-limit?type=design&node-id=337-7972&mode=design&t=iXRryaVFKXuvyjcT-4)  

Global version: [How to upsell 5 mask limit](https://www.figma.com/file/i0SNijb8Twcn7czbzcDLYj/How-to-upsell-5-mask-limit?type=design&node-id=337-8443&mode=design&t=iXRryaVFKXuvyjcT-4)

# Checklist (Definition of Done)
- [ ] Product Owner accepted the User Story (demo of functionality completed) or waived the privilege.
- [ ] All acceptance criteria are met.
- [ ] Jira ticket has been updated (if needed) to match changes made during the development process.
- [ ] I've added or updated relevant docs in the docs/ directory
- [ ] Jira ticket has been updated (if needed) with suggestions for QA when this PR is deployed to stage.
- [ ] All UI revisions follow the [coding standards](https://github.com/mozilla/fx-private-relay/blob/main/docs/coding-standards.md), and use Protocol tokens where applicable (see `/frontend/src/styles/tokens.scss`).
- [ ] Commits in this PR are minimal and [have descriptive commit messages](https://chris.beams.io/posts/git-commit/).
- [ ] l10n changes have been submitted to the l10n repository, if any.

### CODE REVIEW

**Comment by lloan** [comment]
Pending: Upgrade to premium link should be instrumented in GA (e.g., “upgrade-premium-header-mask-limit”)

File: frontend/pendingTranslations.ftl [hunk 1]
```diff
@@ -21,3 +21,14 @@ faq-question-availability-answer-v4 = Free { -brand-name-relay } is available in
...
 modal-custom-alias-picker-form-prefix-label-3 = Enter the text that goes before the @ symbol:
+
+# Mask limit has been reached. 
+# Variables:
+# $limit (number) - the current limit of free masks you can create.
+profile-maxed-aliases-tooltip = You’ve used all { $limit } masks. You can reuse them, but using a unique mask for each account is the most secure option.
+profile-maxed-aliases-with-phone-header = Maximize your email and phone protection
+profile-maxed-aliases-without-phone-header = Maximize your { -brand-name-relay } email protection
+profile-maxed-aliases-with-phone-description = You’ve hit your { $limit }-mask limit on your free { -brand-name-relay } account. Upgrade to { -brand-name-relay-premium } for unlimited email masks, plus the ability to protect your phone number.
+profile-maxed-aliases-without-phone-description = You’ve hit your { $limit }-mask limit on your free { -brand-name-relay } account. Upgrade to { -brand-name-relay-premium } for unlimited email masks.
```
[the remark this answers is not in the document]
Comment (bcolsson) [reply frontend/pendingTranslations.ftl:32]: @pdehaan  - thanks for the tag. I had myself added to a couple pages for an ongoing migration with bedrock. Normally @peiying2 looks after Relay, so I'll have her respond as this is her project.

File: frontend/src/pages/accounts/profile.page.tsx [hunk 2]
```diff
@@ -202,6 +207,9 @@ const Profile: NextPage = () => {
     }
   };
 
+  const freeMaskLimit = 5;
+  const freeMaskLimitReached = allAliases.length >= 5 && !profile.has_premium;
```
[the remark this answers is not in the document]
Comment (lloan) [reply frontend/src/pages/accounts/profile.page.tsx:211]: Great catch! I went ahead and implemented that.

File: frontend/pendingTranslations.ftl (hunk 1 above)
[the remark this answers is not in the document]
Comment (peiying2) [reply frontend/pendingTranslations.ftl:32]: Thank you Bryan for bringing this to my attention. 

@lloan typically I conduct my reviews in the l10n repo, not here unless by special request. Bryan has been helping me to migrate Relay content to mozilla.org, thus his involvement here. He should be asked to review the PRs here. Thank you.

File: frontend/src/pages/accounts/profile.module.scss [hunk 3]
```diff
@@ -197,6 +249,53 @@
...
+    @media screen and #{$mq-lg} {
+      flex-direction: row;
+      gap: $spacing-xs;
+    }
+  }
+
+  .upsell-banner-header {
+    @include text-title-md;
+    font-weight: 700;
+  }
```
Comment (pdehaan) [inline frontend/src/pages/accounts/profile.module.scss:278]: nit: move this closer to the `.upsell-banner-header` definition on L256?

File: frontend/src/pages/accounts/profile.module.scss [hunk 4]
```diff
@@ -197,6 +249,53 @@
...
+
+  .upsell-banner-image {
+    max-width: $content-md;
+    width: 100%;
+  }
+
+  .upsell-banner-description {
+    @include text-body-lg;
+  }
```
Comment (pdehaan) [inline frontend/src/pages/accounts/profile.module.scss:287]: nit: move this closer to the `.upsell-banner-description` definition on L255?

**Comment by pdehaan** [comment]
@lloan I'm seeing the following error(s) in the `test_frontend` check:

```sh
FAIL src/pages/accounts/profile.test.tsx
  ● Test suite failed to run

    A jest worker process (pid=2766) was terminated by another process: signal=SIGKILL, exitCode=null. Operating system logs may contain more information on why this occurred.

      at ChildProcessWorker._onExit (../node_modules/jest-worker/build/workers/ChildProcessWorker.js:370:23)

...

Jest: "global" coverage threshold for statements (50%) not met: 49.11%
Jest: "global" coverage threshold for branches (50%) not met: 37.78%
Jest: "global" coverage threshold for lines (50%) not met: 48.88%
Test Suites: 1 failed, 19 passed, 20 total
Tests:       2 skipped, 150 passed, 152 total
Snapshots:   0 total
Time:        123.481 s
Ran all test suites.
npm ERR! Lifecycle script `test` failed with error: 
npm ERR! Error: command failed 
npm ERR!   in workspace: frontend@0.1.0 
npm ERR!   at location: /home/circleci/project/frontend 

Exited with code exit status 1
```

**Comment by lloan** [comment]
> @lloan I'm seeing the following error(s) in the `test_frontend` check:
> 
> ```shell
> FAIL src/pages/accounts/profile.test.tsx
>   ● Test suite failed to run
> 
>     A jest worker process (pid=2766) was terminated by another process: signal=SIGKILL, exitCode=null. Operating system logs may contain more information on why this occurred.
> 
>       at ChildProcessWorker._onExit (../node_modules/jest-worker/build/workers/ChildProcessWorker.js:370:23)
> 
> ...
> 
> Jest: "global" coverage threshold for statements (50%) not met: 49.11%
> Jest: "global" coverage threshold for branches (50%) not met: 37.78%
> Jest: "global" coverage threshold for lines (50%) not met: 48.88%
> Test Suites: 1 failed, 19 passed, 20 total
> Tests:       2 skipped, 150 passed, 152 total
> Snapshots:   0 total
> Time:        123.481 s
> Ran all test suites.
> npm ERR! Lifecycle script `test` failed with error: 
> npm ERR! Error: command failed 
> npm ERR!   in workspace: frontend@0.1.0 
> npm ERR!   at location: /home/circleci/project/frontend 
> 
> Exited with code exit status 1
> ```

Yes, there are two failing tests that have to do with banners. Going to continue working on this today to figure out why. Thanks.

File: frontend/src/pages/accounts/profile.module.scss [hunk 5]
```diff
@@ -71,12 +71,60 @@
...
+            // so limiting the tooltip to its wrapper's size makes it too thin:
+            max-width: unset;
+          }
+
+          &::before {
+            content: "";
+            height: 2 * $arrowWidth;
+            width: 2 * $arrowWidth;
+            position: absolute;
+            top: -1 * $arrowWidth;
```
Comment (flozia) [inline frontend/src/pages/accounts/profile.module.scss:108]: Question: Wouldn’t these need to be wrapped in `calc()`?

File: frontend/src/pages/accounts/profile.page.tsx [hunk 6]
```diff
@@ -373,6 +423,47 @@ const Profile: NextPage = () => {
   const topBanners = allAliases.length > 0 ? banners : null;
   const bottomBanners = allAliases.length === 0 ? banners : null;
 
+  // Render the upsell banner when a user has reached the free mask limit
+  const UpsellBanner = () =>
+    freeMaskLimitReached && isPhonesAvailableInCountry(runtimeData.data) ? (
```
Comment (flozia) [inline frontend/src/pages/accounts/profile.page.tsx:428]: Nit: Instead of a ternary operator and explicitly returning `null` we could just use `&&` which will also return `null` if the condition evaluates to `false` …

```suggestion
    freeMaskLimitReached && isPhonesAvailableInCountry(runtimeData.data) && (    
```
… and then replace `) : null;` with just `);` on `L465`.

File: frontend/src/pages/accounts/profile.test.tsx [hunk 7]
```diff
@@ -340,6 +338,7 @@ describe("The dashboard", () => {
       configurable: true,
     });
     render(<Profile />);
+    setMockMinViewportWidth(true);
```
Comment (flozia) [inline frontend/src/pages/accounts/profile.test.tsx:341]: Question: This might not be needed since we are not setting `setMockMinViewportWidth(false)` before in this test?
```suggestion
```

**Review by flozia — APPROVED** [review]
Approving for this PR to get unblocked after the test failures are now resolved. Looks good to me and works well — just left some minor comments inline.

Maybe @codemist or @Vinnl could give the changes a quick glance as well since I’m not too familiar with Relay.

File: frontend/src/pages/accounts/profile.page.tsx [hunk 8]
```diff
@@ -213,7 +226,7 @@ const Profile: NextPage = () => {
     ) : (
       <>
         <a className={styles["open-button"]} href="#mpp-choose-subdomain">
-          {l10n.getString("profile-label-set-your-custom-domain")}
+          {l10n.getString("profile-label-set-your-custom-domain-free-user")}
```
Comment (Vinnl) [inline frontend/src/pages/accounts/profile.page.tsx:229]: Just a reminder to mark the old string as `# Deprecated` in the l10n PR.

File: frontend/src/pages/accounts/profile.test.tsx [hunk 9]
```diff
@@ -266,7 +263,7 @@ describe("The dashboard", () => {
       name: '[<Localized> with id [modal-email-domain-available] and vars: {"custom_domain_full":"spongebob.mozmail.com"}]',
     });
     expect(subdomainDialog).toBeInTheDocument();
-  });
+  }, 10000);
```
Comment (Vinnl) [inline frontend/src/pages/accounts/profile.test.tsx:266]: Why was this added? I don't see any particular reason why this test would need more time to execute than others?

File: frontend/src/pages/accounts/profile.test.tsx [hunk 10]
```diff
@@ -328,7 +326,7 @@ describe("The dashboard", () => {
 
     expect(chromeExtensionBanner).not.toBeInTheDocument();
     expect(firefoxBanner).toBeInTheDocument();
-  });
+  }, 10000);
```
Comment (Vinnl) [inline frontend/src/pages/accounts/profile.test.tsx:329]: Same here - this doesn't look like it should be necessary?

File: frontend/src/pages/accounts/profile.test.tsx [hunk 11]
```diff
@@ -835,7 +835,7 @@ describe("The dashboard", () => {
 
     const customAlias = screen.queryByText(/address3/);
     expect(customAlias).toBeInTheDocument();
-  });
+  }, 10000);
```
Comment (Vinnl) [inline frontend/src/pages/accounts/profile.test.tsx:838]: And here too, not clear to me why this should be necessary?

File: frontend/src/pages/accounts/profile.page.tsx [hunk 12]
```diff
@@ -373,6 +423,47 @@ const Profile: NextPage = () => {
...
+              href="/premium#pricing"
+              ref={headerBannerSubscriptionLinkRef}
+            >
+              {l10n.getString("profile-maxed-aliases-cta")}
+            </LinkButton>
+          </div>
+          <Image
+            className={styles["upsell-banner-image"]}
+            src={
+              isPhonesAvailableInCountry(runtimeData.data)
```
Comment (Vinnl) [inline frontend/src/pages/accounts/profile.page.tsx:457]: Nice job using this function rather than just checking if the user in the US as the story suggests :)

File: frontend/src/pages/accounts/profile.page.tsx [hunk 13]
```diff
@@ -255,9 +303,19 @@ const Profile: NextPage = () => {
...
-              {numberFormatter.format(allAliases.length)}
-            </dd>
+            {allAliases.length >= freeMaskLimit &&
+            !profile.has_premium &&
+            isPhonesAvailableInCountry(runtimeData.data) ? (
+              <MaxedMasksTooltip>
+                <dd className={`${styles.value} ${styles.maxed}`}>
+                  {numberFormatter.format(allAliases.length)}
+                </dd>
+              </MaxedMasksTooltip>
```
Comment (Vinnl) [inline frontend/src/pages/accounts/profile.page.tsx:313]: I think this will [result in invalid HTML](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dl):

> **Permitted content** Either: Zero or more groups each consisting of one or more `<dt>` elements followed by one or more `<dd>` elements, optionally intermixed with           `<script>` and `<template>` elements. Or: (in WHATWG HTML, W3C HTML 5.2           and later) One or more `<div>` elements, optionally intermixed with `<script>` and           `<template>` elements.

Would it be possible to swap them around?

```suggestion
              <dd className={`${styles.value} ${styles.maxed}`}>
                <MaxedMasksTooltip>
                  {numberFormatter.format(allAliases.length)}
                </MaxedMasksTooltip>
              </dd>
```

File: frontend/src/pages/accounts/profile.module.scss [hunk 14]
```diff
@@ -71,12 +71,60 @@
...
+          top: calc(100% + $arrowWidth + $spacing-xs);
+          right: $spacing-xs;
+          background-color: $color-purple-80;
+          box-shadow: $box-shadow-sm;
+          display: flex;
+          flex-direction: column;
+          gap: $spacing-sm;
+          padding: $spacing-md $spacing-md;
+          border-radius: $border-radius-sm;
+          // Prevent other .stat-wrapper elements from overlapping this tooltip:
```
Comment (Vinnl) [inline frontend/src/pages/accounts/profile.module.scss:94]: :heart:

File: frontend/src/pages/accounts/profile.module.scss (hunk 5 above)
Comment (Vinnl) [reply frontend/src/pages/accounts/profile.module.scss:108]: This calculation can be done at build time (`$arrowWidth` is a Sass variable, and thus never changes at runtime), so Sass will do it at build time. In other words, this is SCSS syntax, not CSS :)

**Review by Vinnl — COMMENTED** [review]
It looks awesome from the screenshot, and I don't have anything particularly big to point out from reading the code. However, I'm unfortunately unable to try it out myself; the moment I create my fifth mask in [the preview deployment](https://deploy-preview-3748--fx-relay-demo.netlify.app/accounts/profile/?mockId=empty), I get a client-side application error ([this one](https://legacy.reactjs.org/docs/error-decoder.html/?invariant=185)).

Additionally, from your screenshot, and the fact that I don't see any changes in the `<ProfileBanners>` component, I don't think this criterion has been implemented?

> Do not show any other banners at the top of the experience (including if they don’t have the add-on)

And for bonus points, you could add a unit test that verifies that the banner is shown (and others are hidden) in the right circumstances :)

File: frontend/src/pages/accounts/profile.test.tsx (hunk 9 above)
Comment (lloan) [reply frontend/src/pages/accounts/profile.test.tsx:266]: Flaky test - sometimes it fails. There is an underlying issue that should be addressed. There were two like this that were blocking other PRs as well. Might be something we want to look at once we're all back from AH.

File: frontend/src/pages/accounts/profile.test.tsx (hunk 10 above)
Comment (lloan) [reply frontend/src/pages/accounts/profile.test.tsx:329]: https://github.com/mozilla/fx-private-relay/pull/3748#discussion_r1294644891

File: frontend/src/pages/accounts/profile.test.tsx (hunk 11 above)
Comment (lloan) [reply frontend/src/pages/accounts/profile.test.tsx:838]: https://github.com/mozilla/fx-private-relay/pull/3748#discussion_r1294644891

File: frontend/src/pages/accounts/profile.test.tsx (hunk 9 above)
Comment (Vinnl) [reply frontend/src/pages/accounts/profile.test.tsx:266]: I came across this PR by @jwhitlock: https://github.com/mozilla/fx-private-relay/pull/3782 - does that solve the issue? If so, maybe all these `, 10000` could be removed again?

**Comment by lloan** [comment]
@Vinnl added follow up issue for test: https://github.com/mozilla/fx-private-relay/issues/3797 - handled everything else.
